pkg/space/space.go

```diff
@@ -29,6 +32,42 @@ func NewSpace(userSignup *toolchainv1alpha1.UserSignup, targetClusterName string
 	return space
 }
 
+// NewSpaceWithFeatureToggles is the same as NewSpace() but also does a feature toggle lottery drawing
+// and adds the corresponding feature annotations for features which "won" and should be enabled for the space.
+func NewSpaceWithFeatureToggles(userSignup *toolchainv1alpha1.UserSignup, targetClusterName string, compliantUserName, tier string, config toolchainconfig.ToolchainConfig) *toolchainv1alpha1.Space {
+	s := NewSpace(userSignup, targetClusterName, compliantUserName, tier)
+	addFeatureToggles(s, config)
+	return s
+}
+
+// addFeatureToggles does "lottery" drawing for all feature toggles defined in the configuration according to their weights.
+// And it adds the corresponding feature annotation to the space for features that won and should be enabled for the space.
+func addFeatureToggles(space *toolchainv1alpha1.Space, config toolchainconfig.ToolchainConfig) {
+	toggles := config.Tiers().FeatureToggles()
+	var winners string
+	for _, t := range toggles {
+		weight := int(t.Weight())
+		// We generate a random number between 0 and 100. If the number is equal to or lower than the weight
+		// then the feature wins.
+		// We don't use recommended crypto/rand here because we don't need crypto grade random generator
+		// and math/rand with pseudo random numbers, which is much cheaper, is sufficient, so we are disabling the linter
+		if weight == 100 || (weight > 0 && rand.Intn(100) <= weight) { // nolint:gosec
```

Comment:
For better testability, we could modify this `addFeatureToggles` to accept a function that produces random numbers.

```suggestion
// NewSpaceWithFeatureToggles is the same as NewSpace() but also does a feature toggle lottery drawing
// and adds the corresponding feature annotations for features which "won" and should be enabled for the space.
func NewSpaceWithFeatureToggles(
	userSignup *toolchainv1alpha1.
	targetClusterName string,
	compliantUserName, tier string,
	config toolchainconfig.ToolchainConfig,
) *toolchainv1alpha1.Space {
	// We don't use recommended crypto/rand here because we don't need crypto grade random generator
	// and math/rand with pseudo random numbers, which is much cheaper, is sufficient, so we are disabling the linter
	return newSpaceWithFeatureTogglesAndRandomizer(userSignup, targetClusterName, compliantUserName, config, rand.Intn)  // nolint:gosec
}

func newSpaceWithFeatureToggles(
	userSignup *toolchainv1alpha1.UserSignup,
	targetClusterName string,
	compliantUserName, tier string,
	config toolchainconfig.ToolchainConfig, 
	randIntn func(maxValue int) int,
) *toolchainv1alpha1.Space {
	s := NewSpace(userSignup, targetClusterName, compliantUserName, tier)
	addFeatureToggles(s, config, randIntn)
	return s
}

// addFeatureToggles does "lottery" drawing for all feature toggles defined in the configuration according to their weights.
// And it adds the corresponding feature annotation to the space for features that won and should be enabled for the space.
func addFeatureToggles(space *toolchainv1alpha1.Space, config toolchainconfig.ToolchainConfig, randIntn func(maxValue int) int) {
	toggles := config.Tiers().FeatureToggles()
	var winners string
	for _, t := range toggles {
		weight := int(t.Weight())
		// We generate a random number between 0 and 100. If the number is equal to or lower than the weight
		// then the feature wins.
		if weight == 100 || (weight > 0 && randIntn(100) <= weight) {
```

With this change, in tests we will provide a mock function and test the behaviors we are interested in:
* `weight = 0`
* `weight = 100`
* `weight < randIntn(100)`
* `weight == randIntn(100)`
* `weight > randIntn(100)`

Reply:
+1 nice idea!

Reply:
But in this case the test coverage would be lower because we are not testing the default random number generator which would be used in production. Unless I'm missing something.
What is the benefit of such approach?

Reply:
Yeah. you're right. We won't test the code running in production in these unit tests and coverage will be lower.

However, for unit tests that may be good. To me here we should be mostly interested into testing that given different values for `weight` and `randIntn` the logic behave as we expect. So long we ensure that `addFeatureToggles` is using the values generated from `randIntn func(maxValue int) int` as expected, and we know that `rand.Intn` is generating numbers following a distribution that's good for us, I'd say we are ok. 
Alternatively, we may fall into testing that `rand.Intn` is returning values that respects some criteria (distribution) we define, instead of testing the `addFeatureToggles` logic itself, and we may risk to have flaky tests.

Using the approach proposed here, we'll be more precise into testing our logic and we'll left out only a wrapper that's configuring our logic to use the go standard random number generator.  Indeed, we'll be able to test the `addFeatureToggles` logic independently from implementation details about how we generate the random numbers and we'll be able to test with precision the scenarios in which `weight < randIntn(100)`, `weight == randIntn(100)`, and `weight > randIntn(100)`.

I'm trying to think to a valid approach for testing the default behavior (`addFeatureToggles` + `math/rand.Intn`), but I'm not having any good idea.

Reply:
Maybe we can implement the `mySoSoFeature` test to cover `addFeatureToggles` + `math/rand.Intn`? I mean this test case [here](https://github.com/codeready-toolchain/host-operator/blob/012f0cdf58e4fe53e37a2122ef5eb176009527ad/pkg/space/space_test.go#L96)

Reply:
But if we make a mistake and instead of passing something else instead of `rand.Intn()` to that function in the runtime code then this test won't catch it.
So I still believe the current approach is superior because it gives a better result: catches more potential problems. And the risk of a flaky test here is close to 0.5^100 * 2. Which realistically speaking means 0.

Reply:
That's good for me. Btw, if I'm not mistaken, refactoring we'll be able to test the behavior with and without `rand.Intn`, so we should be able to explicitly test the same behaviors we're already testing plus 3 more test cases (somehow implicitly tested if we do not refactor). So, when it comes to the completeness of tests, I feel like it's mostly equivalent.

That said, it seems to me that the difference between the two approaches is not that big, and I'd say we can just choose the one that we prefer without any harm.